### ISSUE TITLE
Add instance at cursor

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#435]
**Describe the project you are working on:**
2D game with lots of different custom nodes.

**Describe the problem or limitation you are having in your project:**
I instance scenes a lot. Well, this is pretty common, lots of users asked for "scene tiles". But I don't even use TileMaps, so I instance nodes manually anyways. I even bound a shortcut to Instance Child Scene to quickly spawn scenes. My problem is that the instances all always created at the origin. You can of course drag and drop from file system at desired position, but due to sheer amount of different scenes and the way my files are organized, this is non-viable. It's much easier for me to press shortcut, fuzzy-search the scene name and press Enter. But then I need to move the node from (0, 0) to positions like (1200, 4600), which wastes my time.

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Would be cool if we had an option to instance a scene at the cursor position. Could be with Ctrl + Click or maybe then Selection List could be utilized (like, showing Add Instance on top of node list, dunno). 

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
![image](https://user-images.githubusercontent.com/2223172/73676104-08910200-46b4-11ea-9f8d-3cf7f7d9a383.png)
![image](https://user-images.githubusercontent.com/2223172/73676117-15adf100-46b4-11ea-82d0-b5a950a3c64b.png)
![image](https://user-images.githubusercontent.com/2223172/73676151-29f1ee00-46b4-11ea-91bf-fe453a058b71.png)

**Describe implementation detail for your proposal (in code), if possible:**
It's basically Instance Child Scene, but called from 2D viewport and positioning the node at cursor.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
It _might_ be possible to replicate the node selection dialog and add a CanvasEditor plugin to do that, but I imagine it would be really cumbersome. Although if the quick open dialog was exposed to plugins... Still, that's more than few lines.

**Is there a reason why this should be core and not an add-on in the asset library?:**
Not sure how many people have similar problem, but I can't be the only one (_;\_;_). And it's a probably small QoL thing using already existing functionality in a way impossible for custom plugins.

**Comment by Zireael07** [comment]
+1. I am working on a space game at a realistic scale, so you can guess stuff's placed really far from origin.
Requesting similar thing for 3D, too. The distances involved aren't quite as large, but still, having to move stuff manually from origin when you have a playing field of 500x500 metres or so is a pain.

**Comment by Jummit** [comment]
> but due to sheer amount of different scenes and the way my files are organized, this is non-viable.

You can favorite them, that's what I do when I need to instance a scene often.

**Comment by KoBeWi** [comment]
> You can favorite them

And have hundreds of favorites? Well, I could only favorite the scenes I need _right now_, but there are some scenes I need pretty much always and I'd have to favorite/unfavorite other scenes a lot. That sounds really messy.

**Comment by Jummit** [comment]
Yeah, favorites get really clumsy when you have more than ten, but if you only have a few it works really great.

**Comment by groud** [comment]
Maybe we could implement a "paint mode" ? Where you could select a scene and instantiate it in one click on the viewport ? That could help in games where you have a lot of collectible to place on the screen.

**Comment by KoBeWi** [comment]
> Maybe we could implement a "paint mode" ? Where you could select a scene and instantiate it in one click on the viewport ?

Could be useful, but we'd also need a preview of where the instance will be placed (obviously).

Also I just realized this could be useful for built-in nodes, not only instances. I just had to create a Node2D and move it very long distance :I (with no other node to use reparenting trick). I'll update the OP.

**Comment by willnationsdev** [comment]
@groud 

> Maybe we could implement a "paint mode" ? Where you could select a scene and instantiate it in one click on the viewport ?

I feel like this and some sort of easy-to-use grid placing of elements, complete with a layer system, would effectively fix a lot of people's problems with TileMap/GridMap by migrating the features people expect out of those nodes to the engine as a whole. I could easily imagine having a more Tiled or GameMaker: Studio like experience of painting instances into the world in specific grid locations (toggleable versus world space) and being able to create layers of elements *without* having to parent them under a specific node. Editor-level node groups would go a long way towards that since you'd be able to filter by group. There's already a WIP PR for adding editor-time global groups too.

**Comment by KoBeWi** [comment]
> Although I never used favorites besides favorite nodes, I'll try it at least...

After some time of using favorites, I started to get lost in the list (17 right now). The problem is that they are really messy. There are some scenes I instance less frequently, but I still need them now and then. Also I often add something new, or add something temporarily, because I need it now and then almost not anymore. It makes it hard to sort the favorites, also it became hard to distinguish them and find the one I look for (some custom icons would be helpful there).

tl;dr I just confirmed that using favorites is not a viable alternative

**Comment by mnemoli** [comment]
I've started a paint mode plugin for 3D. I know there are existing 'scatter' plugins but I believe what people are looking for here is a simple single-object click-to-place functionality. https://github.com/mnemoli/godotplacer (There are a lot of things I would do differently if this were integrated into the engine code - for example, I'd like it to be an actual _tool_, but it looks like the tool set is not currently extendable via plugins.)

Unity and UE4 do not come with this functionality built-in. I think it'd really give Godot a productivity edge if we did have something like this in the main engine.

**Comment by chucklepie** [comment]
An alternative to right-click (not saying any way is better) would be to provide this via the dialog, this way there is still just one way to add a scene instance and less work. It's less fine grained but once placed you will always be moving it regardless.

![image](https://user-images.githubusercontent.com/12863685/106458006-b0880980-6487-11eb-9a4f-b58014ed05fd.png)

**Comment by KoBeWi** [comment]
Except I already implemented it and been using it for months in a custom build: https://github.com/godotengine/godot/pull/41437
It proved to me to be very handy and with grid snapping enabled you can easily instance the node exactly where you want.

**Comment by chucklepie** [comment]
Which stable release will this make it in?

What would be really nice is this same feature but allow to move selected
node on scene inspector to this position also 😁

**Comment by KoBeWi** [comment]
3.2.4 won't get new features anymore, so this won't be added until 3.2.5 at least (if 3.2 goes that far).

**Comment by KoBeWi** [comment]
Implemented for 2D with https://github.com/godotengine/godot/pull/41437
The 3D counterpart could be done in similar way probably.

**Comment by Dadaskis** [comment]
@Calinou written a suggestion in the closed issue mentioned below. I think it deserves to be here:

> Creating a node at the current camera position may be confusing. Instead, it may be better to queue node creation until you click somewhere, and create the node at the clicked location (using a raycast to determine where to place the node exactly).

**Comment by me2beats** [comment]
I'm for having it as a keyboard shortcuts as well.
And an option in Editor Settings to disable this for RMB, because I often move right mouse to rotate the viewport camera

**Comment by Dadaskis** [comment]
I think this is the best place to share some code snippets for editor plugins, if you don't mind.

In this code, a new node is created. Here it is just "Spatial" but i think you can add more types quite easily. If you want - we can discuss it, but for now here it is, the bare-minimum editor plugin code that allows you to create a node in a view space, with a little bonus such as raycast-check that will place the node on the hit position if it hits anything.

It is written for Godot 3.5.1. and works on that one version. This is a hacky way, but worth sharing anyway.

```
tool
extends EditorPlugin

var camera: Camera
var current_scene

# Let's assume that somehow you call this one function
# through beautiful dock UI or something like that
func create_node():
	var node = Spatial.new()

	current_scene.add_child(node)
	node.owner = current_scene 
	# ^^^ Make sure to set owner to make it show up in editor

	# Just saying...
	# This way you can check if the node you are creating has transforms
	# This is useful if you are creating plugin that may create the ones
	# without transforms like WorldEnvironment and others
	#if node.get("transform") == null:
	#	return

	# Just doing a relatively simple raycast check
	# If it hits the surface, it will place the node right at the hit position
	# If it's not, placing somewhere in view space
	var space_state = \
		node.get_world().direct_space_state as PhysicsDirectSpaceState
	var forward = -camera.global_transform.basis.z
	var ray_start = camera.global_transform.origin
	var ray_end = ray_start + (forward * 10)
	var ray = space_state.intersect_ray(ray_start, ray_end)
	if ray:
		node.global_transform.origin = ray.position
	else:
		node.global_transform.origin = camera.global_transform.origin
		node.global_transform.origin += (-camera.global_transform.basis.z * 3.0)
		node.global_transform.origin += (-camera.global_transform.basis.y * 1.0)
	
	# That's it, you have "Spatial" placed near your view.
	# This is really, really minimalistic recreation of logic that my own
	# plugin is using, i hope it's going to be helpful for others.

func forward_spatial_gui_input(camera, event):
	camera = camera
	current_scene = get_tree().get_edited_scene_root()
	return false

func handles(node):
	if node.is_class("Resource"): 
		# Ignoring resources because it can cause deadly lag.
		return false
	return true

func _enter_tree():
	set_input_event_forwarding_always_enabled() 
```
